README.md
# EffToxDesign: EffTox Phase I/II Clinical Trial Design

R package implementation of an adaptive Bayesian design for dose-finding in combined phase I/II clinical trials based on trade-offs between the probabilities of treatment efficacy and toxicity, as developed by Thall and Cook (2004, Biometrics 60, 684-693).


## Requirements

Use of the package requires that a C++ compiler be installed, such as the one provided by the [Rtools](https://cran.r-project.org/bin/windows/Rtools/) software collection available for Windows.


## Installation

The package can be installed directly from GitHub.  First, install the devtools package in R if not done so already.

```R
install.packages("devtools")
```

Then, execute the following command.

```R
devtools::install_github("brian-j-smith/EffToxDesign")
```

## Usage

```R
library(EffToxDesign)
```

## Example

### Trial Design Parameters

An ``EffToxDesign`` R6 class and method functions are provided for EffTrox trial design.  Designs are specified with the ``new`` method function.

```R
doses <- c(1, 2, 4, 6.6, 10)
design <- EffToxDesign$new(
  # Treatment doses
  doses = doses,

  # Acceptable efficacy and toxicity parameters
  piE = 0.5,
  pEL = 0.1,
  piT = 0.3,
  pTL = 0.1,

  # Points defining the utility function
  pi1E = 0.5,
  pi2T = 0.65,
  pi3E = 0.7,
  pi3T = 0.25,
  
  # Logit model prior hyperparameters
  thetaE_mean = efftox_theta(c(0.2, 0.4, 0.6, 0.8, 0.90), doses, 2),
  thetaE_sd = c(2.5423, 2.4406, 0.2),
  thetaT_mean = efftox_theta(c(0.02, 0.04, 0.06, 0.08, 0.10), doses),
  thetaT_sd = c(3.5487, 3.5018),
  psi_mean = 0,
  psi_sd = 1,
  
  # Starting dose
  starting_dose = doses[2],
  
  # Cohort sizes
  cohort_sizes = rep(3, 13)
)
```

Once created, characteristics of the design can be obtained, including effective samples sizes of the prior specifications and utility contours.

```R
design$ESS()

      eff       tox 
0.8305102 0.6750093 

design$contour()
```

### Simulation Studies

Trials can be simulated for a design in order to estimate its operating characteristics.  With use of the *doParallel* package, simulations can be run in parallel to shorten runtimes.

```R
library(doParallel)
registerDoParallel(cores = 6)

sim <- design$simulate(n = 10,
                       eff = c(0.20, 0.40, 0.60, 0.80, 0.90),
                       tox = c(0.05, 0.10, 0.15, 0.20, 0.40),
                       seed = 123)
sim$summary()

                 1         2          4       6.6        10  all
Pr(eff)  0.2000000 0.4000000  0.6000000 0.8000000 0.9000000   NA
Pr(tox)  0.0500000 0.1000000  0.1500000 0.2000000 0.4000000   NA
selected 0.0000000 0.1000000  0.4000000 0.2000000 0.3000000  1.0
given    0.1230769 0.1692308  0.4076923 0.1538462 0.1461538  1.0
E(n_eff) 1.3000000 2.6000000  9.5000000 4.9000000 5.1000000 23.4
E(n_tox) 0.1000000 0.8000000  2.9000000 1.3000000 1.5000000  6.6
E(n)     4.8000000 6.6000000 15.9000000 6.0000000 5.7000000 39.0

sim$barplot()
sim$barplot(stat = c("E(n_tox)", "E(n)"))
```

### Dose-Transition Pathways

Dose-transition pathways show all possible outcomes (E = efficacy only, T = toxicity only, B = both, N = neither) and the doses that would result from n = 1 or more subsequent cohorts.

```R
design$dtp(n = 1, seed = 123)

   cohort1 cohort2
1     2BBB       1
2     2BBE       1
3     2BBN       1
4     2BBT       1
5     2BEE       1
6     2BEN       1
7     2BET       1
8     2BNN       1
9     2BNT       1
10    2BTT       1
11    2EEE       2
12    2EEN       3
13    2EET       1
14    2ENN       3
15    2ENT       1
16    2ETT       1
17    2NNN       3
18    2NNT       3
19    2NTT      NA
20    2TTT      NA
```

### Trial Conduct

Efficacy and toxicity events can be added to the design as they are observed, and dose selection performed on the events.

```R
design$add(yE = c(0, 1, 0), yT = c(0, 0, 0), doses = c(2, 2, 2))
selected <- design$select(seed = 123)
selected$dose
selected$plot()
selected$superiority()
```

Events can be removed from the design altogether with the ``reset`` method or selectively with the ``drop`` and ``keep`` methods (see documenation).

```R
design$reset()
```


NEWS.md
# News

## Version Updates

## 0.4
* Fix compilation of rstan shared library.

## 0.3
* Added EffToxDesign option to suppress log transformation and centering of doses.
* Added EffToxDesign as.stan and size methods.
* Added EffToxSelect plot method for posterior density and dose-response curves.

## 0.2

* Added EffToxSelect density plots for efficacy and toxicity probabilities.
* Added EffToxSim and TwoPhaseSim bar plots for summary statistics.
* Renamed EffToxSim and TwoPhaseSim summary statistics "E(eff)" and "E(tox)" to "E(n_eff)" and "E(n_tox)".
* Added citation information.

## 0.1

* Initial package release


R/EffTox.R
# Part of the EffToxDesign package for estimating model parameters
# Copyright (C) 2018 Brian J Smith
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#' @title Calculate the p-index for EffTox utility contours
#'
#' @description
#' Calculate the p-index for EffTox utility contours so that the neutral utility
#' contour intersects the following points in the
#' Prob(Efficacy) - Prob(Toxicity) plane:
#' (\code{pi1E}, 0), (1, \code{pi2T}) and (\code{pi3E}, \code{pi3T})
#'
#' @param pi1E Efficacy probability required when toxicity is impossible;
#' a number between 0 and 1
#' @param pi2T Toxicity probability permitted when efficacy is guaranteed;
#' a number between 0 and 1
#' @param pi3E Efficacy probability of an equi-utility third point
#' @param pi3T Toxicity probability of an equi-utility third point
#'
#' @return The p-index
#'
#' @examples
#' efftox_solve_p(0.5, 0.65, 0.7, 0.25)
#'
#' @references Thall, Herrick, Nguyen, Venier & Norris. 2014, Effective sample
#' size for computing prior hyperparameters in Bayesian phase I-II dose-finding
#'
#' @noRd
#'
efftox_solve_p <- function(pi1E, pi2T, pi3E, pi3T) {
  # Calculate p for the efficacy/toxicity contours that will intersect points
  # (pi1E, 0), (eff.star, tox.star), and (1, pi2T)
  
  .objective = function(p, pi1E, pi2T, pi3E, pi3T) {
    a <- ((1 - pi3E) / (1 - pi1E))
    b <- pi3T / pi2T
    return(a^p + b^p - 1)
  }
  
  rt <- stats::uniroot(.objective, interval = c(0, 100),
                       pi1E = pi1E, pi2T = pi2T, pi3E = pi3E,
                       pi3T = pi3T)
  return(rt$root)
}


#' @title Estimate EffTox logit model coefficients
#'
#' @description
#' Estimate regression coefficients for the efficacy or toxicity logit models in
#' an EffTox trial design, given a set of probabilities and doses.
#'
#' @param probs vector of probabilites at which to estimate coefficients.
#' @param doses vector of corresponding doses.
#' @param degree degree of the polynomial dose effect in the logit model.
#' @param log logical indicating whether doses are to be log transformed and
#' centered in the analysis.
#'
#' @return A vector of coefficients.
#'
#' @seealso
#' \code{\link{EffToxDesign}}
#'
efftox_theta <- function(probs, doses, degree = 1, log = TRUE) {
  if(log) doses <- logcenter(doses)
  y <- log(probs / (1 - probs))
  structure(lm(y ~ poly(doses, degree, raw = TRUE))$coef, names = NULL)
}


#' @title Get the utility of efficacy & toxicity probability pairs
#'
#' @description Get the utility of efficacy & toxicity probability pairs
#'
#' @param eff Probability of efficacy; number between 0 and 1
#' @param tox Probability of toxicity; number between 0 and 1
#' @param p p-index of EffTox utility contours. Use \code{efftox_solve_p}
#' @param pi1E Efficacy probability required when toxicity is impossible;
#' a number between 0 and 1
#' @param pi2T Toxicity probability permitted when efficacy is guaranteed;
#' a number between 0 and 1
#'
#' @return Utility value(s)
#'
#' @examples
#' p <- efftox_solve_p(0.5, 0.65, 0.7, 0.25)
#'
#' u <- efftox_utility(eff = 0.7, tox = 0.25, p, 0.5, 0.65)
#' round(u, 4) == 0
#'
#' u <- efftox_utility(eff = c(0.6, 0.7, 0.8), tox = c(0.1, 0.2, 0.3), p, 0.5, 0.65)
#' round(u, 2) == c(0.04, 0.08, 0.12)
#'
#' @seealso \code{\link{efftox_solve_p}}
#'
#' @noRd
#'
efftox_utility <- function(eff, tox, p, pi1E, pi2T) {
  a <- ((1 - eff) / (1 - pi1E))
  b <- tox / pi2T
  r <- (a^p + b^p)^(1/p)
  1 - r
}


#' @title Parse a string of EffTox outcomes to binary vector notation.
#'
#' @description Parse a string of EffTox outcomes to the binary vector notation
#' required by Stan for model invocation. The outcome string describes the doses
#' given and outcomes observed. The format of the string is described in Brock
#' et al. (2017). The letters E, T, N and B are used to represents patients that
#' experienced (E)fficacy only, (T)oxicity only, (B)oth efficacy and toxicity,
#' and (B)oth. These letters are concatenated after numerical dose-levels to
#' convey the outcomes of cohorts of patients. For instance, \code{2ETB}
#' represents a cohort of three patients that were treated at dose-level 2, and
#' experienced efficacy, toxicity and both events, respectively. The results of
#' cohorts are separated by spaces. Thus, \code{2ETB 1NN} extends our previous
#' example, where the next cohort of two were treated at dose-level 1 and both
#' patients experienced neither efficacy nor toxicity. See examples.
#'
#' We present the notation in the EffTox setting but it is applicable in
#' general seamless phase I/II dose-finding scenarios.
#'
#' @param outcome_string character string, conveying doses given and outcomes
#' observed.
#'
#' @return a list with elements \code{yE}, \code{yT}, \code{levels} and
#' \code{n}. These elements are congruent with the same in
#' \code{efftox_params}.
#'
#' @examples
#' x = efftox_parse_outcomes('1NNE 2EEN 3TBB')
#' x$n == 9
#' x$yE == c(0, 0, 1, 1, 1, 0, 0, 1, 1)
#' sum(x$yT) == 3
#'
#' @references Brock et al. (submitted 2017), Implementing the EffTox
#' Dose-Finding Design in the Matchpoint Trial.
#'
#' @noRd
#'
efftox_parse_outcomes <- function(outcome_string) {
  regex = '\\w*\\d+[ETBN]+\\w*'
  outcomes = outcome_string
  m = gregexpr(pattern = regex, text = outcomes)
  cohorts = regmatches(outcomes, m)
  if(length(cohorts) <= 0) stop(paste('Matching', outcomes,
                                      'to EffTox outcomes failed'))
  cohorts = cohorts[[1]]
  levels = c(); yE = c(); yT = c();
  n = 0
  for(cohort in cohorts) {
    # print(cohort)
    mc = regexpr('(\\d+)', cohort)
    dl_start = mc[1]
    dl_end = mc[1] + attr(mc,"match.length") - 1
    dl = substr(cohort, dl_start, dl_end)
    dl = as.integer(dl)
    outcomes = substr(cohort, dl_end + 1, nchar(cohort))
    outcomes = strsplit(outcomes, "")[[1]]
    these_doses = rep(dl, length(outcomes))
    levels = c(levels, these_doses)
    these_eff = as.integer((outcomes == 'E') | (outcomes == 'B'))
    yE =c(yE, these_eff)
    these_tox = as.integer((outcomes == 'T') | (outcomes == 'B'))
    yT = c(yT, these_tox)
    n = n + length(outcomes)
  }
  return(list(
    levels = levels, yE = yE, yT = yT, n = n
  ))
}
